src/main/java/io/aiven/kafka/auth/audit/Auditor.java

```diff
@@ -0,0 +1,139 @@
+/**
+ * Copyright (c) 2020 Aiven, Helsinki, Finland. https://aiven.io/
+ */
+
+package io.aiven.kafka.auth.audit;
+
+import java.io.Serializable;
+import java.net.InetAddress;
+import java.time.format.DateTimeFormatter;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.StringJoiner;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.kafka.common.security.auth.KafkaPrincipal;
+
+import kafka.network.RequestChannel.Session;
+import kafka.security.auth.Operation;
+import kafka.security.auth.Resource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class Auditor {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(Auditor.class);
+
+    protected final Map<AuditKey, UserActivity> audit = new HashMap<>();
+
+    private final ScheduledExecutorService auditScheduler =
+        Executors.newScheduledThreadPool(1);
+
+    private final long eventsAggregationPeriod;
+
+    public Auditor(final long eventsAggregationPeriod) {
+        this.eventsAggregationPeriod = eventsAggregationPeriod;
+    }
+
+    public void addActivity(final Session session,
+                            final Operation operation,
+                            final Resource resource,
+                            final Boolean hasAccess) {
+
+        final AuditKey auditKey = new AuditKey(session.principal(), session.clientAddress());
+        synchronized (audit) {
+            audit.compute(auditKey, (key, userActivity) -> {
+                if (Objects.isNull(userActivity)) {
+                    return onUserActivity(new UserActivity(), operation, resource, hasAccess);
+                } else {
+                    return onUserActivity(userActivity, operation, resource, hasAccess);
+                }
+            });
+        }
+    }
+
+    public void start() {
+        auditScheduler.scheduleAtFixedRate(
+            this::dump,
+            eventsAggregationPeriod,
+            eventsAggregationPeriod,
+            TimeUnit.SECONDS
+        );
+    }
+
+    public void stop() {
+        auditScheduler.shutdownNow();
+        dump();
+    }
+
+    protected void dump() {
+        synchronized (audit) {
+            audit.forEach((key, userActivity) -> LOGGER.info("{}", auditMessage(key, userActivity)));
+            audit.clear();
+        }
```

Comment:
This lock is on a pretty hot path of user requests, so I'd be great to reduce time spent in it as much as possible. A possible way would be to substitute the current audit map with a fresh one, and then dump the data from the old map outside of the lock, like drafted [here](https://github.com/aiven/aiven-kafka-auth/pull/6#discussion_r403048265).